### PR TITLE
[Security] [Core] String utils refactor

### PR DESCRIPTION
**Pull request by ircmaxell**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR |  |

This supersedes #13984 (it includes it, but also includes additional refactoring).

Since length information is leaked in any case, preventing unnecessary duplication of secrets is important. Since casting will _always_ make a copy, we only cast if absolutely necessary. Additionally, appending will create a new copy of the secret, so we avoid doing that, but instead only iterate over the minimum of the two strings.

### CODE REVIEW

**Comment by sarciszewski** [comment]
:+1: looks great to me

File: src/Symfony/Component/Security/Core/Util/StringUtils.php [hunk 1]
```diff
@@ -38,29 +38,55 @@ private function __construct()
...
         }
 
-        $knownLen = strlen($knownString);
-        $userLen = strlen($userInput);
+        $knownLen = self::safeStrlen($knownString);
+        $userLen = self::safeStrlen($userInput);
 
-        // Extend the known string to avoid uninitialized string offsets
-        $knownString .= $userInput;
+		if ($userLen != $knownLen) {
```
Comment (hhamon) [inline src/Symfony/Component/Security/Core/Util/StringUtils.php]: Wrong indent?

File: src/Symfony/Component/Security/Core/Util/StringUtils.php [hunk 2]
```diff
@@ -38,29 +38,55 @@ private function __construct()
...
+        for ($i = 0; $i < $knownLen; $i++) {
             $result |= (ord($knownString[$i]) ^ ord($userInput[$i]));
         }
 
         // They are only identical strings if $result is exactly 0...
         return 0 === $result;
     }
+
+    /**
+     * Return the number of bytes in a string
```
Comment (hhamon) [inline src/Symfony/Component/Security/Core/Util/StringUtils.php]: `Returns the number of bytes in a string.`

File: src/Symfony/Component/Security/Core/Util/StringUtils.php [hunk 3]
```diff
@@ -38,29 +38,55 @@ private function __construct()
...
+     *
+     * @param string $string The string whose length we wish to obtain
+     * @return int
+     */
+    public static function safeStrlen($string)
+    {
+        // Premature optimization
+        // Since this cannot be changed at runtime, we can cache it
+        static $func_exists = null;
+        if ($func_exists === null) {
```
Comment (hhamon) [inline src/Symfony/Component/Security/Core/Util/StringUtils.php]: `if (null === $funcExists) {`

File: src/Symfony/Component/Security/Core/Util/StringUtils.php [hunk 4]
```diff
@@ -38,29 +38,55 @@ private function __construct()
...
+     * Return the number of bytes in a string
+     *
+     * @param string $string The string whose length we wish to obtain
+     * @return int
+     */
+    public static function safeStrlen($string)
+    {
+        // Premature optimization
+        // Since this cannot be changed at runtime, we can cache it
+        static $func_exists = null;
```
Comment (hhamon) [inline src/Symfony/Component/Security/Core/Util/StringUtils.php]: `static $funcExists = null;`

File: src/Symfony/Component/Security/Core/Util/StringUtils.php [hunk 5]
```diff
@@ -38,29 +38,55 @@ private function __construct()
...
+        if ($func_exists === null) {
+            $func_exists = function_exists('mb_strlen');
```
Comment (hhamon) [inline src/Symfony/Component/Security/Core/Util/StringUtils.php]: `$funcExists = function_exists('mb_strlen');`

File: src/Symfony/Component/Security/Core/Util/StringUtils.php [hunk 6]
```diff
@@ -38,29 +38,55 @@ private function __construct()
...
+        }
+
+        if ($func_exists) {
```
Comment (hhamon) [inline src/Symfony/Component/Security/Core/Util/StringUtils.php]: `if ($funcExists) {`

**Comment by fabpot** [comment]
Closing in favor of #14024, which includes these changes (it's much easier for everyone to work on one PR).

**Comment by ircmaxell** [comment]
This makes me sad.

**Comment by ircmaxell** [comment]
@hhamon thanks for the review. If this ever gets re-opened I'll handle the issues. But seeing as it's closed, there's no point.

File: src/Symfony/Component/Security/Core/Util/StringUtils.php (hunk 1 above)
Comment (jakzal) [inline src/Symfony/Component/Security/Core/Util/StringUtils.php]: We also prefer strong comparison `!==`.

**Comment by fabpot** [comment]
@sarciszewski Sorry about this. I thought the third PR was really just about getting the changes from your PR and #14028, and then adding on top of it, but I read too fast (just by looking at the commit list.) I apologize for the mistake and I'm reopening this PR.

I'm no security expert and I trust you and @ircmaxell on those matters, so I will let you let me know which one(s) should be merged.

**Comment by fabpot** [comment]
I've merged this PR in the 2.3 branch (I've fixed the CS in 2c67400). Thank you @ircmaxell and sorry for the confusion yesterday.

**Comment by sarciszewski** [comment]
:+1: glad to see this resolved in days rather than months :)

**Pull request closed without merging**